## iHerb parser

In [1]:
import requests as req  #request library
import pandas as pd
import numpy as np
import time

In [2]:
!pip install fake_useragent #a library for generating a fake user agent

ERROR: Invalid requirement: '#a'


In [3]:
!pip install beautifulsoup4  #library for structuring the data of the primary html code of the page

ERROR: Invalid requirement: '#library'


In [4]:
!pip install lxml

In [5]:
from fake_useragent import UserAgent # let's load one of the methods of fake_useragent library
import bs4
from bs4 import BeautifulSoup as bs

In [6]:
#the main page link
page_link = 'https://il.iherb.com/c/eye-vision'

#making a request
response = req.get(page_link, headers={'User-Agent': UserAgent().chrome, 'referer':'https://www.google.com/'})
response

<Response [200]>

In [7]:
html=response.content

len(html)

1273493

In [8]:
html[:500]

b'\n\n<!DOCTYPE html>\n\n\n\n\n<!--[if IE 9 ]> <html class="ie9 he list productlist &#x5E2;&#x5D9;&#x5E0;&#x5D9;&#x5D9;&#x5DD;-&#x5D5;&#x5E8;&#x5D0;&#x5D9;&#x5D9;&#x5D4;-list" dir="rtl" lang="he"> <![endif]-->\n<!--[if IE 10 ]> <html class="ie10 he list productlist &#x5E2;&#x5D9;&#x5E0;&#x5D9;&#x5D9;&#x5DD;-&#x5D5;&#x5E8;&#x5D0;&#x5D9;&#x5D9;&#x5D4;-list" dir="rtl" lang="he"> <![endif]-->\n<!--[if (gt IE 9)|!(IE)]><!-->\n\n\n<html class="ih-public ih-catalog cc-il he list productlist category &#x5E2;&#x5D9;&#'

In [9]:
soup=bs(html, 'html.parser')    #lxml

type(soup)

bs4.BeautifulSoup

In [10]:
print(soup.prettify()[:1000])

<!DOCTYPE html>
<!--[if IE 9 ]> <html class="ie9 he list productlist &#x5E2;&#x5D9;&#x5E0;&#x5D9;&#x5D9;&#x5DD;-&#x5D5;&#x5E8;&#x5D0;&#x5D9;&#x5D9;&#x5D4;-list" dir="rtl" lang="he"> <![endif]-->
<!--[if IE 10 ]> <html class="ie10 he list productlist &#x5E2;&#x5D9;&#x5E0;&#x5D9;&#x5D9;&#x5DD;-&#x5D5;&#x5E8;&#x5D0;&#x5D9;&#x5D9;&#x5D4;-list" dir="rtl" lang="he"> <![endif]-->
<!--[if (gt IE 9)|!(IE)]><!-->
<html class="ih-public ih-catalog cc-il he list productlist category עיניים-וראייה-list core-app" dir="rtl" lang="he">
 <!--<![endif]-->
 <head>
  <!-- production/catalog/bahrain 2055 -->
  <link href="https://s3.images-iherb.com/" rel="dns-prefetch"/>
  <link href="https://s3.images-iherb.com/static/i/favicon-iherb/favicon.ico" rel="icon" type="image/ico">
   <link href="https://il.iherb.com/c/eye-vision" rel="canonical"/>
   <link href="https://zw.iherb.com/c/eye-vision" hreflang="en-ZW" rel="alternate"/>
   <link href="https://zm.iherb.com/c/eye-vision" hreflang="en-ZM" rel="alternat

In [11]:
soup.html.head.title

<title> עיניים וראייה </title>

In [12]:
type(soup.html.head.title)

bs4.element.Tag

In [13]:
#pulling out the text
soup.html.head.title.text

' עיניים וראייה '

In [14]:
#finding an element by class
obj = soup.find('a', attrs = {'class':'absolute-link product-link'})
obj

<a aria-label="California Gold Nutrition, Immune 4, Immune System Support, 60 Veggie Capsules" class="absolute-link product-link" data-ga-brand-code="CGN" data-ga-discount-price="₪14.74" data-ga-event="click" data-ga-event-action="productClick" data-ga-event-category="Ecommerce" data-ga-event-label="Product List_ListView:special" data-ga-product-id="101714" data-ga-product-position="1" data-part-number="CGN-01842" data-product-id="101714" href="https://il.iherb.com/pr/california-gold-nutrition-immune-4-immune-system-support-60-veggie-capsules/101714" itemprop="url" title="California Gold Nutrition, Immune 4, Immune System Support, 60 Veggie Capsules"></a>

In [15]:
obj.get('class')

['absolute-link', 'product-link']

In [16]:
type(obj)

bs4.element.Tag

In [17]:
#link to our object
obj.attrs['href']

'https://il.iherb.com/pr/california-gold-nutrition-immune-4-immune-system-support-60-veggie-capsules/101714'

In [18]:
type(obj.attrs['href'])

str

In [19]:
#finding all the elements with 1 address
vitamine_links=soup.findAll(lambda tag: tag.name == 'a' and tag.get('class') == ['absolute-link', 'product-link'])
vitamine_links[:3]

[<a aria-label="California Gold Nutrition, Immune 4, Immune System Support, 60 Veggie Capsules" class="absolute-link product-link" data-ga-brand-code="CGN" data-ga-discount-price="₪14.74" data-ga-event="click" data-ga-event-action="productClick" data-ga-event-category="Ecommerce" data-ga-event-label="Product List_ListView:special" data-ga-product-id="101714" data-ga-product-position="1" data-part-number="CGN-01842" data-product-id="101714" href="https://il.iherb.com/pr/california-gold-nutrition-immune-4-immune-system-support-60-veggie-capsules/101714" itemprop="url" title="California Gold Nutrition, Immune 4, Immune System Support, 60 Veggie Capsules"></a>,
 <a aria-label="California Gold Nutrition, Immune 4, Immune System Support, 180 Veggie Capsules" class="absolute-link product-link" data-ga-brand-code="CGN" data-ga-discount-price="₪42.13" data-ga-event="click" data-ga-event-action="productClick" data-ga-event-category="Ecommerce" data-ga-event-label="Product List_ListView:special" 

In [20]:
len(vitamine_links)

24

In [21]:
### cleansing links
vitamine_links= [link.attrs['href'] for link in vitamine_links]

In [22]:
vitamine_links[:5]

['https://il.iherb.com/pr/california-gold-nutrition-immune-4-immune-system-support-60-veggie-capsules/101714',
 'https://il.iherb.com/pr/california-gold-nutrition-immune-4-immune-system-support-180-veggie-capsules/101716',
 'https://il.iherb.com/pr/california-gold-nutrition-astaxanthin-astalif-pure-icelandic-12-mg-120-veggie-softgels/71684',
 'https://il.iherb.com/pr/natural-factors-eye-factors-with-2-mg-lutein-90-capsules/2612',
 'https://il.iherb.com/pr/california-gold-nutrition-omega-3-premium-fish-oil-180-epa-120-dha-100-fish-gelatin-softgels/62118']

In [23]:
len(vitamine_links)

24

In [24]:
#a function that returns a list of product links received from the current page
def getPageLinks(page_number):
    
    page_link = 'https://il.iherb.com/c/eye-vision?noi=24&p={}'.format(page_number)  #link to the search page
    
    response = req.get(page_link, headers={'User-Agent': UserAgent().chrome})  #requesting data on this page
    
    if not response.ok:
        
        return []          #if the server refused us, we will return an empty sheet for the current page
    
    html = response.content    #getting the content of the page and translating it into soup
    soup = bs(html,'html.parser')
    
    #looking for links to products and clearing them of unnecessary tags
    vitamine_links=soup.findAll(lambda tag:tag.name=='a' and tag.get('class') == ['absolute-link', 'product-link'])
    vitamine_links = [link.attrs['href'] for link in vitamine_links]
    
    return vitamine_links

In [25]:
getPageLinks(30)

['https://il.iherb.com/pr/natural-factors-omegafactors-ultra-prim-evening-primrose-oil-1000-mg-180-softgels/2605',
 'https://il.iherb.com/pr/solgar-flaxseed-oil-1-250-mg-250-softgels/9033',
 'https://il.iherb.com/pr/source-naturals-zeaxanthin-with-lutein-10-mg-60-capsules/4547',
 'https://il.iherb.com/pr/source-naturals-provinal-omega-7-30-softgels/55415',
 'https://il.iherb.com/pr/organic-traditions-sprouted-chia-flax-seed-powder-16-oz-454-g/106041',
 'https://il.iherb.com/pr/zenwise-health-marine-algae-derived-vegan-omega-3-225-mg-60-softgels/105705',
 'https://il.iherb.com/pr/doctor-s-best-resveratrol-french-red-wine-grape-extract-60-mg-90-veggie-caps/18',
 'https://il.iherb.com/pr/nordic-naturals-ultimate-omega-2x-strawberry-560-mg-60-mini-soft-gels/22346',
 'https://il.iherb.com/pr/thorne-research-r-lipoic-acid-60-capsules/18591',
 'https://il.iherb.com/pr/wiley-s-finest-wild-alaskan-fish-oil-vitamin-k2-650-mg-60-fish-softgels/73697',
 'https://il.iherb.com/pr/solaray-pumpkin-seed

For 1 vitamine product:

In [26]:
vitamine_page = 'https://il.iherb.com/pr/sports-research-omega-3-fish-oil-triple-strength-1-250-mg-90-softgels/90284'

response = req.get(vitamine_page, headers={'User-Agent': UserAgent().chrome, 'referer':'https://www.google.com/'})
html = response.content

soup = bs(html, 'lxml')

In [27]:
#pull out the price of the product
price = soup.find('div', attrs={'class':'col-xs-15 col-md-15 price our-price'}) 
price

<div class="col-xs-15 col-md-15 price our-price" id="price">
₪85.88                                    </div>

In [28]:
price = price.get_text()
price

'\n₪85.88                                    '

In [29]:
price=price.split()[0]
price

'₪85.88'

In [30]:
price = price.replace('₪', '')
price

'85.88'

We can do the same for the brand, name, description, rating, review, quantity

In [31]:
brand=soup.find('span', attrs={'itemprop': 'name'})
brand

<span itemprop="name">
<bdi>Sports Research</bdi>
</span>

In [32]:
brand = brand.get_text()
brand

'\nSports Research\n'

In [33]:
name = soup.find('div',attrs={'class':'product-summary-title', 'itemprop':'name', 'id':'name'})
name

<div class="product-summary-title" id="name" itemprop="name">
        Sports Research‏, שמן דגים אומגה 3, עוצמה משולשת, 1,250 מ"ג, 90 כמוסות רכות
    </div>

In [34]:
name=name.get_text()
name

'\n        Sports Research\u200f, שמן דגים אומגה 3, עוצמה משולשת, 1,250 מ"ג, 90 כמוסות רכות\n    '

In [35]:
description=soup.find('div', attrs={'itemprop':'description'})
description

<div itemprop="description"><ul> <li>מיוצר עם AlaskOmega מזהבון בר אלסקי</li> <li>שמן דגים מאושר על ידי IFOS (תכנית תקנים בינלאומית לשמן דגים) בדירוג של 5 כוכבים</li> <li>תוסף תזונה</li> <li>עשוי ללא רכיבים מהונדסים גנטית</li> <li>ללא גלוטן</li> <li>תעודת GMP‏ (תנאי ייצור נאותים)</li> <li>אושר ונבדק על-ידי צד שלישי</li> <li>אושר על ידי MSC (המועצה לשימור ימי) כמאכל ים בר קיימא</li> <li>זיקוק מולקולרי מתקדם ומזוכך</li> <li>מתעלה על הארגון העולמי ל-EPA (חומצה איקוסאפנטאנואית) ו-DHA (חומצה דוקוסהקסנואית)</li> <li>100% מקורי ומיוצר בארצות הברית</li> <li>כמוסה רכה אחת של שמן דגים אומגה-3 של SR* = 3 כמוסות רכות סטנדרטיות של שמן דגים</li></ul><p>שמן דגים אומגה-3 של Sports Research מכיל ריכוזים חזקים במיוחד של EPA ו-DHA בכל כמוסה של ג'לטין דגים (937 מ"ג). זהבון אלסקי שמקורו במים קרים, שנתפס בר (לא מדגים שגדלו בחווה).</p><p>*כל כמוסת אומגה-3 של Sports Research מכילה סך הכל 80% חומצות שומן חיוניות (EFAs) מסוג אומגה-3.</p></div>

In [36]:
description = description.get_text()
description

' מיוצר עם AlaskOmega מזהבון בר אלסקי שמן דגים מאושר על ידי IFOS (תכנית תקנים בינלאומית לשמן דגים) בדירוג של 5 כוכבים תוסף תזונה עשוי ללא רכיבים מהונדסים גנטית ללא גלוטן תעודת GMP\u200f (תנאי ייצור נאותים) אושר ונבדק על-ידי צד שלישי אושר על ידי MSC (המועצה לשימור ימי) כמאכל ים בר קיימא זיקוק מולקולרי מתקדם ומזוכך מתעלה על הארגון העולמי ל-EPA (חומצה איקוסאפנטאנואית) ו-DHA (חומצה דוקוסהקסנואית) 100% מקורי ומיוצר בארצות הברית כמוסה רכה אחת של שמן דגים אומגה-3 של SR* = 3 כמוסות רכות סטנדרטיות של שמן דגיםשמן דגים אומגה-3 של Sports Research מכיל ריכוזים חזקים במיוחד של EPA ו-DHA בכל כמוסה של ג\'לטין דגים (937 מ"ג). זהבון אלסקי שמקורו במים קרים, שנתפס בר (לא מדגים שגדלו בחווה).*כל כמוסת אומגה-3 של Sports Research מכילה סך הכל 80% חומצות שומן חיוניות (EFAs) מסוג אומגה-3.'

In [37]:
rating=soup.find('a', attrs={'class':'rating-count'})
rating

<a class="rating-count" data-ds-target="link" href="https://il.iherb.com/r/sports-research-omega-3-fish-oil-triple-strength-1-250-mg-90-softgels/90284" title="4.7/5 - 33740 ביקורות"><span>33740</span> ביקורות</a>

In [38]:
rating = rating.get('title')
rating

'4.7/5 - 33740 ביקורות'

In [39]:
review = rating.split()[2]
review

'33740'

In [40]:
rating = rating.split()[0]
rating = rating[0:3]
rating

'4.7'

In [41]:
quantity=soup.find('span', attrs={'class':'current-item'})
quantity

<span class="current-item" data-current-item="90 Count">90 Count</span>

In [42]:
quantity = quantity.get_text()
quantity

'90 Count'

In [43]:
import re
quantity = re.findall(r'\d+', quantity)
quantity

['90']

Let's write functions to get different statistics

In [44]:
def getStats(soup, stats):

    try:
        obj = soup.find('div', attrs={'class':'col-xs-15 col-md-15 price our-price', 'id':'price'})
        obj = obj.get_text()
        obj = obj.split()[0]
        obj = obj.replace('₪', '')
    except:
        obj=None
    
    return obj

In [45]:
def getStats2(soup, stats):

    try:
        obj = soup.find('span', attrs={'itemprop': 'name'})
        obj = obj.get_text()
        obj = obj.replace('\n', '')
    except:
        obj=None
    
    return obj

In [46]:
def getStats3(soup, stats):

    try:
        obj = soup.find('div', attrs={'itemprop':'description'})
        obj = obj.get_text()
    except:
        obj=None
    
    return obj

In [47]:
def getStats4(soup, stats):

    try:
        obj = soup.find('div',attrs={'class':'product-summary-title', 'itemprop':'name', 'id':'name'})
        obj = obj.get_text()
        obj = obj.replace('\n', '')
    except:
                          
        obj=None
    
    return obj

In [48]:
def getStats5(soup, stats):

    try:
        obj = soup.find('a', attrs={'class':'rating-count'})
        obj = obj.get('title')
        obj = obj.split()[0]
        obj = obj[0:3]
    except:
        obj=None
        
    return obj

In [49]:
def getStats6(soup, stats):

    try:
        obj = soup.find('a', attrs={'class':'rating-count'})
        obj = obj.get('title')
        obj = obj.split()[2]
    except:
        obj=None
        
    return obj

In [50]:
def getStats7(soup, stats):

    try:
        obj = soup.find('span', attrs={'class':'current-item'})
        obj = obj.get_text()
        obj = obj.split()[0]                        
    except:
        obj=None
        
    return obj

In [51]:
price = getStats(soup, stats='price')
brand = getStats2(soup, stats='brand')
description = getStats3(soup, stats='description')
name = getStats4(soup, stats='name')
rating = getStats5(soup, stats='rating')
review = getStats6(soup, stats='review')
quantity = getStats7(soup, stats='quantity')


print("Price: {}\nName:{}\nBrand:{}\nQuantity: {}\nDescription: {}\nRating: {}\nReview: {}".format(price, name, brand, quantity, description, rating, review))

Price: 85.88
Name:        Sports Research‏, שמן דגים אומגה 3, עוצמה משולשת, 1,250 מ"ג, 90 כמוסות רכות    
Brand:Sports Research
Quantity: 90
Description:  מיוצר עם AlaskOmega מזהבון בר אלסקי שמן דגים מאושר על ידי IFOS (תכנית תקנים בינלאומית לשמן דגים) בדירוג של 5 כוכבים תוסף תזונה עשוי ללא רכיבים מהונדסים גנטית ללא גלוטן תעודת GMP‏ (תנאי ייצור נאותים) אושר ונבדק על-ידי צד שלישי אושר על ידי MSC (המועצה לשימור ימי) כמאכל ים בר קיימא זיקוק מולקולרי מתקדם ומזוכך מתעלה על הארגון העולמי ל-EPA (חומצה איקוסאפנטאנואית) ו-DHA (חומצה דוקוסהקסנואית) 100% מקורי ומיוצר בארצות הברית כמוסה רכה אחת של שמן דגים אומגה-3 של SR* = 3 כמוסות רכות סטנדרטיות של שמן דגיםשמן דגים אומגה-3 של Sports Research מכיל ריכוזים חזקים במיוחד של EPA ו-DHA בכל כמוסה של ג'לטין דגים (937 מ"ג). זהבון אלסקי שמקורו במים קרים, שנתפס בר (לא מדגים שגדלו בחווה).*כל כמוסת אומגה-3 של Sports Research מכילה סך הכל 80% חומצות שומן חיוניות (EFAs) מסוג אומגה-3.
Rating: 4.7
Review: 33740


Let's write a function that returns all the information on the current product

In [52]:
def getVitamineData(vitamine_page):
    
    response = req.get(vitamine_page, headers={'User-Agent': UserAgent().chrome})
    
    if not response.ok:
        
        return response.status_code
    
    html = response.content
    soup = bs(html,'html.parser')
    
    #using previously written functions, we parse the information
    price = getStats(soup=soup, stats='price')
    brand = getStats2(soup=soup, stats='brand')
    description = getStats3(soup=soup, stats='description')
    name = getStats4(soup=soup, stats='name')
    rating = getStats5(soup=soup, stats='rating')
    review = getStats6(soup=soup, stats='review')
    quantity = getStats7(soup=soup, stats='quantity')
    
    #we are compiling a dictionary in which all the received and processed data will be stored
    data_row = {'price':price, 
                'brand':brand, 
                'description':description, 
                'name':name,
                'rating': rating,
                'review': review,
                'quantity': quantity}
    
    return data_row

In [53]:
data_row=getVitamineData('https://il.iherb.com/pr/sports-research-omega-3-fish-oil-triple-strength-1-250-mg-90-softgels/90284')
data_row

{'price': '85.88',
 'brand': 'Sports Research',
 'description': ' מיוצר עם AlaskOmega מזהבון בר אלסקי שמן דגים מאושר על ידי IFOS (תכנית תקנים בינלאומית לשמן דגים) בדירוג של 5 כוכבים תוסף תזונה עשוי ללא רכיבים מהונדסים גנטית ללא גלוטן תעודת GMP\u200f (תנאי ייצור נאותים) אושר ונבדק על-ידי צד שלישי אושר על ידי MSC (המועצה לשימור ימי) כמאכל ים בר קיימא זיקוק מולקולרי מתקדם ומזוכך מתעלה על הארגון העולמי ל-EPA (חומצה איקוסאפנטאנואית) ו-DHA (חומצה דוקוסהקסנואית) 100% מקורי ומיוצר בארצות הברית כמוסה רכה אחת של שמן דגים אומגה-3 של SR* = 3 כמוסות רכות סטנדרטיות של שמן דגיםשמן דגים אומגה-3 של Sports Research מכיל ריכוזים חזקים במיוחד של EPA ו-DHA בכל כמוסה של ג\'לטין דגים (937 מ"ג). זהבון אלסקי שמקורו במים קרים, שנתפס בר (לא מדגים שגדלו בחווה).*כל כמוסת אומגה-3 של Sports Research מכילה סך הכל 80% חומצות שומן חיוניות (EFAs) מסוג אומגה-3.',
 'name': '        Sports Research\u200f, שמן דגים אומגה 3, עוצמה משולשת, 1,250 מ"ג, 90 כמוסות רכות    ',
 'rating': '4.7',
 'review': '33742',
 'quantity': '90'

Now we can get the final dataframe with the received data

In [54]:
final_df = pd.DataFrame()

In [55]:
final_df = final_df.append(data_row, ignore_index=True)

In [56]:
final_df

brand                                        description  \
0  Sports Research   מיוצר עם AlaskOmega מזהבון בר אלסקי שמן דגים ...   

                                                name  price quantity rating  \
0          Sports Research‏, שמן דגים אומגה 3, עו...  85.88       90    4.7   

  review  
0  33742

In [57]:
from tqdm import tqdm_notebook

In [58]:
vitamine_links

['https://il.iherb.com/pr/california-gold-nutrition-immune-4-immune-system-support-60-veggie-capsules/101714',
 'https://il.iherb.com/pr/california-gold-nutrition-immune-4-immune-system-support-180-veggie-capsules/101716',
 'https://il.iherb.com/pr/california-gold-nutrition-astaxanthin-astalif-pure-icelandic-12-mg-120-veggie-softgels/71684',
 'https://il.iherb.com/pr/natural-factors-eye-factors-with-2-mg-lutein-90-capsules/2612',
 'https://il.iherb.com/pr/california-gold-nutrition-omega-3-premium-fish-oil-180-epa-120-dha-100-fish-gelatin-softgels/62118',
 'https://il.iherb.com/pr/california-gold-nutrition-lutein-with-zeaxanthin-20-mg-60-veggie-softgels/95888',
 'https://il.iherb.com/pr/doctor-s-best-lutein-from-optilut-10-mg-120-veggie-caps/4256',
 'https://il.iherb.com/pr/doctor-s-best-lutein-with-lutemax-2020-20-mg-180-softgels/58903',
 'https://il.iherb.com/pr/california-gold-nutrition-omega-3-premium-fish-oil-180-epa-120-dha-240-fish-gelatin-softgels/86598',
 'https://il.iherb.com/

In [59]:
import tqdm

In [60]:
#let's go through all the products on the page
for vitamine_link in tqdm_notebook(vitamine_links):
    try: 
        data_row = getVitamineData(vitamine_link)
        final_df = final_df.append(data_row, ignore_index=True)
        time.sleep(0.3)
    except:
        continue

C:\Users\Anzhelika Soldatova\anaconda3\lib\site-packages\ipykernel_launcher.py:2: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  


In [61]:
final_df.drop_duplicates(inplace=True)
final_df.shape

(25, 7)

In [62]:
final_df.head(10) 

brand  \
0            Sports Research   
1  California Gold Nutrition   
2  California Gold Nutrition   
3  California Gold Nutrition   
4            Natural Factors   
5  California Gold Nutrition   
6  California Gold Nutrition   
7              Doctor's Best   
8              Doctor's Best   
9  California Gold Nutrition   

                                         description  \
0   מיוצר עם AlaskOmega מזהבון בר אלסקי שמן דגים ...   
1  California Gold Nutrition Immune 4 חיזוק המערכ...   
2  California Gold Nutrition Immune 4חיזוק המערכת...   
3   אסטקסנטין מבית California Gold NutritionAstal...   
4  תוסף תזונהתפקוד תקין של העיןטוהר ועוצמה מובטחי...   
5  California Gold Nutrition®️שמן דגים עם אומגה 3...   
6  לוטאין עם זאקסנטין, 20 מ"ג מבית California Gol...   
7  Science-Based Nutrition™‎מיוצר מ-®OptiLut, תמצ...   
8  Science-Based Nutrition™‎לוטאין & וזאקסנטיןמסי...   
9  California Gold Nutrition®️ שמן דגים עם אומגה ...   

                                                name  price quantity rating  \
0          Sports Research‏, שמן דגים אומגה 3, עו...  85.88       90    4.7   
1          California Gold Nutrition‏, Immune4, ת...   None       60    4.8   
2          California Gold Nutrition‏, Immune 4, ...   None      180    4.8   
3          California Gold Nutrition‏, אסטקסנטין ...   None      120    4.6   
4          Natural Factors‏, תוסף תזונה Eye Facto...   None     None    4.5   
5          California Gold Nutrition‏, אומגה 3, ש...  24.57      100    4.8   
6          California Gold Nutrition‏, לוטאין עם ...  24.57       20    4.7   
7          Doctor's Best‏, לוטאין מ-OptiLut, ‏‏10...  44.03     None    4.5   
8          Doctor's Best‏, לוטאין עם Lutemax 2020...  99.84      180    4.6   
9          California Gold Nutrition‏, אומגה 3, ש...  52.66      240    4.8   

   review  
0   33742  
1   97570  
2   97570  
3   12439  
4    6795  
5  195531  
6   15975  
7   10111  
8    7302  
9  195531

In [63]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25 entries, 0 to 24
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   brand        25 non-null     object
 1   description  25 non-null     object
 2   name         25 non-null     object
 3   price        20 non-null     object
 4   quantity     15 non-null     object
 5   rating       25 non-null     object
 6   review       25 non-null     object
dtypes: object(7)
memory usage: 1.6+ KB


In [64]:
final_df.columns

Index(['brand', 'description', 'name', 'price', 'quantity', 'rating',
       'review'],
      dtype='object')

In [65]:
# final_df = pd.DataFrame(columns=['brand', 'description', 'name', 'price', 'quantity', 'rating', 'review'])
# #let's go through all the pages of products in this category
# for page_number in tqdm.notebook.tqdm(range(80), desc='Pages'):     
    
#     vitamine_links = getPageLinks(page_number) 
#     for vitamine_link in tqdm.notebook.tqdm(vitamine_links, desc='Vitamine', leave=False):
        
#         for i in range(3):
#             try:
                
#                 data_row = getVitamineData(vitamine_link)         
#                 final_df = final_df.append(data_row, ignore_index=True)  
                
#                 break
                
#             except:
                
#                 print('AHTUNG! parsing once again:', vitamine_link)
#                 continue
                

In [66]:
def checkIP():
    ip = req.get('http://checkip.dyndns.org').content
    soup = bs(ip, 'html.parser')
    print(soup.find('body').text)

In [67]:
checkIP()

Current IP Address: 83.130.68.175


In [69]:
final_df = pd.DataFrame(columns=['brand', 'description', 'name', 'price', 'quantity', 'rating', 'review'])

for page_number in tqdm.notebook.tqdm(range(80), desc='Pages'):
    
    vitamine_links = getPageLinks(page_number) 
    for vitamine_link in tqdm.notebook.tqdm(vitamine_links, desc='Vitamine', leave=False):
        
        for i in range(3):
            try:
                
                data_row = getVitamineData(vitamine_link)         
                final_df = final_df.append(data_row, ignore_index=True)  

                break
                
            except:
                
                continue
                
                
    final_df.to_csv('Vitamine_IHerb{}.csv'.format(page_number))
    if page_number//10:
        print(final_df.shape)
final_df.to_csv('Vitamine_IHerb_full.csv')

(240, 7)


(264, 7)


(288, 7)


(312, 7)


(336, 7)


(360, 7)


(384, 7)


(408, 7)


(432, 7)


(456, 7)


(480, 7)


(504, 7)


(528, 7)


(552, 7)


(576, 7)


(600, 7)


(624, 7)


(648, 7)


(672, 7)


(696, 7)


(720, 7)


(744, 7)


(768, 7)


(792, 7)


(816, 7)


(840, 7)


(864, 7)


(888, 7)


(912, 7)


(936, 7)


(960, 7)


(984, 7)


(1008, 7)


(1032, 7)


(1056, 7)


(1080, 7)


(1104, 7)


(1128, 7)


(1152, 7)


(1176, 7)


(1200, 7)


(1224, 7)


(1248, 7)


(1272, 7)


(1296, 7)


(1320, 7)


(1344, 7)


(1368, 7)


(1392, 7)


(1416, 7)


(1440, 7)


(1464, 7)


(1488, 7)


(1512, 7)


(1536, 7)


(1560, 7)


(1584, 7)


(1608, 7)


(1632, 7)


(1656, 7)


(1680, 7)


(1704, 7)


(1728, 7)


(1752, 7)


(1776, 7)


(1800, 7)


(1824, 7)


(1848, 7)


(1872, 7)


(1896, 7)



In [70]:
data=pd.read_csv('Vitamine_IHerb_full.csv')

In [71]:
del data['Unnamed: 0']

In [72]:
data.head()

brand  \
0  California Gold Nutrition   
1  California Gold Nutrition   
2  California Gold Nutrition   
3            Natural Factors   
4  California Gold Nutrition   

                                         description  \
0  California Gold Nutrition Immune 4 חיזוק המערכ...   
1  California Gold Nutrition Immune 4חיזוק המערכת...   
2   אסטקסנטין מבית California Gold NutritionAstal...   
3  תוסף תזונהתפקוד תקין של העיןטוהר ועוצמה מובטחי...   
4  California Gold Nutrition®️שמן דגים עם אומגה 3...   

                                                name  price quantity rating  \
0          California Gold Nutrition‏, Immune4, ת...    NaN       60    4.8   
1          California Gold Nutrition‏, Immune 4, ...    NaN      180    4.8   
2          California Gold Nutrition‏, אסטקסנטין ...    NaN      120    4.6   
3          Natural Factors‏, תוסף תזונה Eye Facto...    NaN      NaN    4.5   
4          California Gold Nutrition‏, אומגה 3, ש...  24.57      100    4.8   

     review  
0   97574.0  
1   97574.0  
2   12439.0  
3    6795.0  
4  195534.0

In [73]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1896 entries, 0 to 1895
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   brand        1896 non-null   object 
 1   description  1896 non-null   object 
 2   name         1896 non-null   object 
 3   price        1699 non-null   float64
 4   quantity     738 non-null    object 
 5   rating       1892 non-null   object 
 6   review       1892 non-null   float64
dtypes: float64(2), object(5)
memory usage: 103.8+ KB


As a result, we received some data on vitamins for eye vision from the pages of iHerb website, it turned out 1896 products from 80 pages, we saved data in a dataframe